### DISCUSSION TITLE
why dotnet?

### DISCUSSION THREAD

**Discussion by davidatkins1** [philosowaffle/peloton-to-garmin#433] [General]
I am wondering why Microsoft .NET is needed to run this on my mac? Seems like for what is being done here...shouldn't need to rely on Microsoft-anything. I can relate...the author on windows machine and thank you for making a mac-runable version; I am just curious what part of the process is dependent on dotnet?

**Comment by philosowaffle** [comment]
Whenever you want to compile and run source code on any computer, that computer needs to understand how to interpret the programming language that code is written in.  The means for interpreting and executing programming languages on a given operating system are often called [runtimes](https://www.techtarget.com/searchsoftwarequality/definition/runtime).

* To run Java code => you need the [Java Runtime installed](https://www.java.com/en/download/manual.jsp)
* To run Python => you need [python installed](https://www.python.org/downloads/) (though many macs do come with it already installed by default)
* To run C# code => you need the .NET Runtime installed.
* Same story for other popular languages like [Go](https://go.dev/doc/install) and [Rust](https://www.rust-lang.org/tools/install) 

Needing to install the .NET runtime should not be confused with anything specifically Windows related, atleast not in how you might traditionally think of Mac vs. Windows.  The .NET Runtime doesn't even come installed by default on Windows machines, which should show how unrelated to Windows it actually is.  While perhaps unfortunately named, the .NET Runtime truly is just there to interpret and run C# code.

I do want to point out that installing the runtime for a language is **only necessary if you plan to compile and run the source code directly.**

For the Windows folks, I provide a pre-compiled `exe` that they can just download and run without having to install any runtime. I could theoretically provide the same for Mac, but there's a reason I don't:

1.  Generally Mac's won't let you install apps that haven't come directly from the App store
2.  I don't want to pay Apple the required $100/year to develop for their App Store
3. I don't want to deal with getting apps approved by Apple, they are notorious for making this process painful, especially for indie devs

So unfortunately, Mac users are stuck with having to compile and run the code themselves, which means their computer needs to understand C#, which means they need the language runtime installed, which happens to be called .NET Runtime.

Hope that helps clarify, there's a lot of layers to this onion :)

**Reply by philosowaffle** [reply]
One layer of the onion I forgot to mention, the C# programming language is owned and developed by Microsoft, that is why you see them using their branding and nomenclature in the runtime name and terminal commands.

**Reply by davidatkins1** [reply]
Oh, that makes sense...for some reason, I had been thinking this was written in python! I don't know why I was thinking that! I didn't realize it was C#

**Reply by philosowaffle** [reply]
Ah! yeah that makes sense, I would be confused too lol You're about 2 years too late on the python front, my original script was in fact in python.  But I'm actually terrible at python, so once this project started gaining popularity I re-wrote it into C#, a language I can much more efficiently support.

**Comment by davidatkins1** [comment]
actually...I first used your application maybe a year or two ago, but then
it broke and I could not get it to work reliably. I was always running it
manually because it would freeze up or something. Tonight I removed all the
old stuff and started over step by step. I think when I last installed it,
there were definitely some python steps including getting the garmin
uploader...at any rate, tonight's install went very smoothly once I 1)
turned on the upload in the config file and 2) got my passwords right.
